### ISSUE TITLE
[BUG] unable to Hide unmapped backend roles

### ISSUE THREAD

**Issue by anubisg1** [opensearch-project/security#2189]
**What is the bug?**
I have linked opensearch to our active directory domain.
I also mapped some AD groups to the proper roles…

everything works but i noticed that Dashboards exposes too much information to the users…
specifically when a user clicks on “view roles and identities” the user can see every Active Directory group the user name is linked to (as backend roles) even if that group is never used in OpenSearch…

this is information i would like to hide as it shows details the user shouldn’t know.
Everything i tried doesn't seem to work

this is the closest thing i found
https://opensearch.org/docs/latest/security-plugin/configuration/ldap/#advanced-exclude-roles-from-nested-role-lookups

my steps in configuration are:

1. Reducing the scope by looking for roles/AD Groups in a "sub" OU.

```
{
  "enable_ssl": false,
  "hosts": [
    "192.168.0.20:389",
    "192.168.0.21:389"
  ],
  "bind_dn": "CN=binduser,OU=Bind Users,OU=UsersOU,DC=customer,DC=company,DC=com",
  "password": "xxxxxx",
  "userbase": "OU=UsersOU,DC=customer,DC=company,DC=com",
  "usersearch": "(sAMAccountName={0})",
  "username_attribute": "sAMAccountName",
  "rolebase": "OU=OpenSearch,OU=User Groups - Customer,DC=customer,DC=company,DC=com",
  "rolesearch": "(uniqueMember={0})"
}
```
this doesn't work as i see the roles also from

`OU=User Groups - Customer,DC=customer,DC=company,DC=com` 

2. configured nested role filter

i added the following configuration
```
                    resolve_nested_roles: true
                    nested_role_filter:
                      - '^(?!.*OpenSearch.*$)'
```
that should match (and therefore filterout) everything except groups which have “OpenSearch” in them

![image](https://user-images.githubusercontent.com/4518931/196626901-a3c1e3c4-5667-42a8-8dfb-c01d42fe2553.png)

but also doesn't work as i still see all the roles.

**What is the expected behavior?**
Only roles used by OpenSearch should be visible to the user as backend-roles
Also, filtering of roles, should work as expected 

**What is your host/environment?**
 - OS: Azure AKS
 - Version - 2.3.0
 - Plugins - nothing manually installed, only default plugins installed by the k8s operator

**Do you have any screenshots?**
![196627548-f3a5a8f8-9284-4e3a-a88a-362cacaf064f](https://user-images.githubusercontent.com/2754967/196763399-51180385-8267-4787-a598-86195717b815.png)

**Comment by peternied** [comment]
@cwperks I've moved the issue, could you update the title/description based on your findings?

**Comment by williamhargrove** [comment]
Was any progress made on this? I am looking to solve exactly the same issue, but seems like this remains outstanding?

**Comment by yoicomment** [comment]
This is a very highly needed feature. When are you planning to add this feat?

**Comment by scrawfor99** [comment]
Hi @yoicomment, I don't think that there is any active work being done on this issue at the moment. I encourage you to open a pull request to implement this feature if you are interested. The maintainers would be happy to review any pull requests around this area. 

Thank you

**Comment by cwperks** [comment]
Would it make sense to add 2 new config options to achieve this? `include_roles` and `exclude_roles`? They would not be able to be configured simultaneously, but if an `include_roles` is configured then only the roles that are included in that list would be returned for the user. If `exclude_roles` is configured then it could return all roles extracted from the LDAP system, excluding the roles in that list.

i.e.

```
authz:
      roles_from_myldap:
        description: "Authorize via LDAP or Active Directory"
        http_enabled: false
        transport_enabled: false
        authorization_backend:
          # LDAP authorization backend (gather roles from a LDAP or Active Directory, you have to configure the above LDAP authentication backend settings too)
          type: ldap
          config:
            enable_ssl: false
            verify_hostnames: true
            hosts:
            - localhost:8389
            rolebase: 'ou=groups,dc=example,dc=com'
            rolesearch: '(member={0})'
            rolename: cn
            userbase: 'ou=people,dc=example,dc=com'
            usersearch: '(uid={0})'
            ...
            include_roles:
              - roleA
              - roleB
              - roleC
```

OR

```
authz:
      roles_from_myldap:
        description: "Authorize via LDAP or Active Directory"
        http_enabled: false
        transport_enabled: false
        authorization_backend:
          # LDAP authorization backend (gather roles from a LDAP or Active Directory, you have to configure the above LDAP authentication backend settings too)
          type: ldap
          config:
            enable_ssl: false
            verify_hostnames: true
            hosts:
            - localhost:8389
            rolebase: 'ou=groups,dc=example,dc=com'
            rolesearch: '(member={0})'
            rolename: cn
            userbase: 'ou=people,dc=example,dc=com'
            usersearch: '(uid={0})'
            ...
            exclude_roles:
              - roleA
              - roleB
              - roleC
```

**Comment by MaciejMierzwa** [comment]
Hi, picked it up

**Comment by MaciejMierzwa** [comment]
Hi @cwperks 
Decided on this version of configuration:
```
authz:
  roles_from_myldap:
    description: "Authorize via LDAP or Active Directory"
    http_enabled: false
    transport_enabled: false
    authorization_backend:
      type: ldap
      config:
        enable_ssl: false
        verify_hostnames: true
        hosts:
        - localhost:8389
        rolebase: 'ou=groups,dc=example,dc=com'
        rolesearch: '(member={0})'
        rolename: cn
        userbase: 'ou=people,dc=example,dc=com'
        usersearch: '(uid={0})'
        ...
        exclude_roles: "(regex)"
          ```
 I think that implementing "include_roles" might change how cluster works, admins would need to change configuration after upgrading. Excluding roles seems to be better, if there is no change to the config cluster will behave the same without interaction.
Here's the PR: https://github.com/opensearch-project/security/pull/3809

**Comment by MaciejMierzwa** [comment]
Hi @anubisg1, @yoicomment, @williamhargrove would this solution work for you?

**Comment by yoicomment** [comment]
So, what would be the next steps now?

**Comment by cwperks** [comment]
The new setting `exclude_roles` will be released in 2.13.0. 

I have opened up a PR in the documentation repo to add this setting to documentation: https://github.com/opensearch-project/documentation-website/pull/6399

Closing this issue as the associated PR and backport to 2.x have been merged.
